sdk/communication/azure-communication-callingserver/src/main/java/com/azure/communication/callingserver/models/ProgressReporter.java

```diff
@@ -0,0 +1,172 @@
+            });
+        }
+    }
+
+    /**
+     * This type is used to keep track of the total amount of data transferred for a single request. This is the type we
+     * will use when the customer uses the factory to add progress reporting to their Flowable. We need this additional
+     * type because we can't keep local state directly as lambdas require captured local variables to be effectively
+     * final.
+     */
+    private static class SequentialProgressReporter extends ProgressReporterImpl {
+        SequentialProgressReporter(ProgressReceiver progressReceiver) {
+            super(progressReceiver);
+        }
+
+        @Override
+        public void reportProgress(long bytesTransferred) {
+            super.reportProgress(bytesTransferred);
+            this.progressReceiver.reportProgress(this.blockProgress);
+        }
+    }
+
+    /**
+     * This type is used to keep track of the total amount of data transferred as a part of a parallel upload in order
+     * to coordinate progress reporting to the end user. We need this additional type because we can't keep local state
+     * directly as lambdas require captured local variables to be effectively final.
+     */
+    private static class ParallelProgressReporter extends ProgressReporterImpl {
+        /*
+        This lock will be instantiated by the operation initiating the whole transfer to coordinate each
+        ProgressReporterImpl.
+         */
+        private final Lock transferLock;
+
+        /*
+        We need an AtomicLong to be able to update the value referenced. Because we are already synchronizing with the
+        lock, we don't incur any additional performance hit here by the synchronization.
+         */
+        private final AtomicLong totalProgress;
+
+        ParallelProgressReporter(ProgressReceiver progressReceiver, Lock lock, AtomicLong totalProgress) {
+            super(progressReceiver);
+            this.transferLock = lock;
+            this.totalProgress = totalProgress;
+        }
+
+        @Override
+        public void reportProgress(long bytesTransferred) {
+            super.reportProgress(bytesTransferred);
+
+            /*
+            It is typically a bad idea to lock around customer code (which the progressReceiver is) because they could
+            never release the lock. However, we have decided that it is sufficiently difficult for them to make their
+            progressReporting code threadsafe that we will take that burden and the ensuing risks. Although it is the
+            case that only one thread is allowed to be in onNext at once, there are multiple independent
+            requests happening at once to stage/download separate chunks, so we still need to lock either way.
+             */
+            transferLock.lock();
+            this.progressReceiver.reportProgress(this.totalProgress.addAndGet(bytesTransferred));
```

Comment:
I took this code from the Storage SDK. In the comment it mentions about this risk:
```
It is typically a bad idea to lock around customer code (which the progressReceiver is) because they could
never release the lock. However, we have decided that it is sufficiently difficult for them to make their
progressReporting code threadsafe that we will take that burden and the ensuing risks. Although it is the
case that only one thread is allowed to be in onNext at once, there are multiple independent
requests happening at once to stage/download separate chunks, so we still need to lock either way.
```